elastic/beats
# [Heartbeat] Add run_once mode in the scheduler

fixes https://github.com/elastic/beats/issues/29280
fixes https://github.com/elastic/beats/issues/29256

## What does this PR do?
This PR fixes data stream mode while running heartbeat in `run_once` mode , before this if you would run heartbeat in run_once mode, it would still push documents to normal heartbeat indexes.

This PR also fixes data like `tags` not being pushed as part of documents.

### Strategy

1. Just use the regular scheduler, not a custom execution path for runOnce
3. Get rid of scheduler.Start figuring out which things to runOnce and how to track them is tricky with state, now any task added to the scheduler runs immediately. This also fixes a bug, where the old implementation would do the same, so things could run before Start previously anyway.
4. Add a new scheduler.WaitForRunOnce method that lets you wait for all Added tasks, it also cleans up for you
5. Refactor the deadline missed reporter in the schedule to properly accommodate RunOnce and also to not be so fragile generally.


## Why is it important?

This is needed for synthetics service where we are using hearting run_once mode.

## Checklist



## How to test this PR locally
Run hearbeat in run once mode using data stream format
for example

you can use following example config, in this branch, data should be pushed to data stream, you can verify that using only 'synthetics-*' index string in the uptime settings page, Alternatively you can verify it using stack management index management to verify data streams for both type of monitors http, browser are created.

On master/main if you will run the same config, it will push data to heartbeat indices. 

```
heartbeat.run_once: true
heartbeat.monitors:
  - type: browser
    id: elastic-monitor
    data_stream:
      namespace: default
    streams:
      - data_stream:
          dataset: browser
          type: synthetics
        id: elastic-monitor
        name: elastic-monitor
        type: browser
        schedule: '@every 60s'
        source:
          inline:
            script: |-
              step('load homepage', async () => {
                await page.goto('https://www.elastic.co');
              });
  - id: 4ae879a9-b5da-4132-94ba-ab6e1fcbdc6e
    name: Sample monitor 2
    type: http
    data_stream:
      namespace: default
    streams:
      - id: synthetics/http-http-4ae879a9-b5da-4132-94ba-ab6e1fcbdc6e
        name: Sample monitor 2
        type: http
        data_stream:
          dataset: http
          type: synthetics
        urls: 'https://elastic.co'
        service.name: APM Service Name
        schedule: '@every 5s'
        timeout: 1600
        max_redirects: 1
        proxy_url: 'https://elastic.co'
        tags:
          - tag
```



## Screenshots

![image](https://user-images.githubusercontent.com/3505601/144829015-14f57a45-e419-42fa-8c18-7347a50b373e.png)
## Logs

Make sure heartbeat ends with "Ending run_once mode" logs
